upstairs/src/downstairs.rs

```diff
@@ -0,0 +1,10875 @@
+                // it better not have already been acked.
+                let job = self.ds_active.get(&ds_id).unwrap();
+                assert!(!job.acked);
+                None
+            }
+        } else {
+            // No live-repair in progress, nothing to do
+            None
+        }
+    }
+
+    /// Pushes live-repair forward, starting at the given job
+    ///
+    /// `self.repair` must be waiting on the job given by `ds_id`, and that job
+    /// must be (1) in `self.ackable_work` and (2) not yet acked.
+    ///
+    /// As such, this function should only be called after
+    /// `self.check_live_repair` provides the value for `ds_id`; they're broken
+    /// into separate functions to avoid locking the `GuestWork` structure if
+    /// live-repair can't continue.
+    ///
+    /// It's possible that handling this job will make subsequent live-repair
+    /// jobs ackable immediately (`test_repair_extent_fail_noop_out_of_order`
+    /// exercises this case).  As such, this function will continue running
+    /// until the next live-repair job is not ready.
+    pub(crate) fn continue_live_repair(
+        &mut self,
+        ds_id: JobId,
+        gw: &mut GuestWork,
+        up_state: &UpstairsState,
+        generation: u64,
+    ) {
+        self.continue_live_repair_inner(ds_id, gw, up_state, generation);
+        while let Some(ds_id) = self.check_live_repair() {
+            self.continue_live_repair_inner(ds_id, gw, up_state, generation);
+        }
+    }
+
+    fn continue_live_repair_inner(
+        &mut self,
+        ds_id: JobId,
+        gw: &mut GuestWork,
+        up_state: &UpstairsState,
+        generation: u64,
+    ) {
+        let done = self.ds_active.get(&ds_id).unwrap();
+        assert!(!done.acked);
+        assert!(self.ackable_work.contains(&ds_id));
+        let r = done.result();
+
+        let Some(repair) = &self.repair else {
+            panic!("cannot continue live-repair without self.repair");
+        };
+
+        match &r {
+            Ok(()) => {
+                // keep going
+                info!(self.log, "got repair ok in {:?}", repair.state);
+            }
+            Err(e) => {
+                error!(self.log, "got repair error {e} in {:?}", repair.state);
+
+                // We keep going here, because we need to submit no-op jobs to
+                // avoid things getting clogged up.
+                if !repair.aborting_repair {
+                    warn!(self.log, "aborting live-repair due to IO error");
+                    self.abort_repair(up_state);
+                }
+            }
+        }
+
+        // It's possible for the Downstairs to have changed state here, if it
+        // disconnected.  We'll check that as well and start aborting the repair
+        // if that's the case.
+        let Some(repair) = &self.repair else {
+            panic!("cannot continue live-repair without self.repair");
```

Comment:
This is a reborrow, but since we already checked that `self.repair` exists above, we can do the more terse `self.repair.as_[ref,mut]().unwrap()`.  Done in 2cf302068f9a94303b31eec4e5df5c4b47a0a503